app/models/katello/activation_key.rb

```diff
@@ -117,12 +117,16 @@ def products
         pools = cp_pools.collect { |cp_pool| Pool.find_pool(cp_pool['id'], cp_pool) }
 
         pools.each do |pool|
-          Product.where(:cp_id => pool.product_id).each do |product|
-            if product.is_a? Katello::MarketingProduct
-              all_products += product.engineering_products
-            else
-              all_products << product
-            end
+          marketing_products = MarketingProduct.includes(:engineering_products, :marketing_engineering_products).
+              where(:cp_id => pool.product_id)
+          products = Product.where(:cp_id => pool.product_id).where('type != ?', "Katello::MarketingProduct")
```

Comment:
instead of doing this for each pool, could you do something like:  where(:cp_id => pools.map(&:product_id)

this way you're not looping over each pool and running 2 queries per pool